pkg/gameservers/health.go

```diff
@@ -126,12 +126,26 @@ func (hc *HealthController) evictedPod(pod *corev1.Pod) bool {
 	return pod.Status.Reason == "Evicted"
 }
 
-// failedContainer checks each container, and determines if there was a failed
-// container
-func (hc *HealthController) failedContainer(pod *corev1.Pod) bool {
+// failedContainerAfterReady checks each container, and determines if there was a failed
+// container after the Gameserver pod has been marked as Ready
+// Since we want restarts to occur pre-Ready state, we only care about post-Ready
+func (hc *HealthController) failedContainerAfterReady(pod *corev1.Pod) bool {
+	containerID, ok := pod.Annotations[agonesv1.GameServerReadyContainerIDAnnotation]
+	// this annotation only exists once the GameServer pod is ready, so if we don't have
+	// it, then it's not after Ready.
+	if !ok {
+		return false
+	}
+
 	container := pod.Annotations[agonesv1.GameServerContainerAnnotation]
 	for _, cs := range pod.Status.ContainerStatuses {
 		if cs.Name == container {
+			// if the current container is the Ready container instance, and is running, then the LastTerminationState
+			// was from before the GameServer was ready, so we can ignore it.
+			if cs.State.Terminated == nil && cs.ContainerID == containerID && cs.LastTerminationState.Terminated != nil {
```

Comment:
Maybe a silly question, but the comment mentions that we can ignore the last termination state but it's still part of the conditional check. If `cs.State.Terminated == nil && cs.ContainerID == containerID` returns true and `cs.LastTerminationState.Terminated != nil` returns false (because this is the first instantiation of the container), then we fail this check. I think that we still correctly return false in the following check, but why not just do it here? Is there an edge cases that this is checking? If so, let's make sure there is a test for it.